kraken_frontend/src/utils/filter/parser.ts

```diff
@@ -382,50 +88,56 @@ function parseAst(input: string, parseColumn: (column: string, cursor: Cursor) =
 }
 
 /** Parse an {@link Expr.Or} expression using a `parseValue` to parse the leaves */
-function parseOr<T>(tokens: Cursor, parseValue: (cursor: Cursor) => Expr.Value<T>): Expr.Or<T> {
-    const list = [parseAnd(tokens, parseValue)];
-    while (tokens.peekToken()?.type === "logicalOr") {
-        tokens.nextToken(); // Consume the `,`
-        list.push(parseAnd(tokens, parseValue));
-    }
-    return { or: list };
+export function parseOr<T>(parseValue: (cursor: Cursor) => Expr.Value<T>): (tokens: Cursor) => Expr.Or<T> {
+    return (tokens: Cursor) => {
+        const list = [parseAnd(parseValue)(tokens)];
+        while (tokens.peekToken()?.type === "logicalOr") {
+            tokens.nextToken(); // Consume the `,`
+            list.push(parseAnd(parseValue)(tokens));
+        }
+        return { or: list };
+    };
 }
 
 /** Parse an {@link Expr.And} expression using a `parseValue` to parse the leaves */
-function parseAnd<T>(tokens: Cursor, parseValue: (cursor: Cursor) => Expr.Value<T>): Expr.And<T> {
-    const list = [parseNot(tokens, parseValue)];
-    while (tokens.peekToken()?.type === "logicalAnd") {
-        tokens.nextToken(); // Consume the `&`
-        list.push(parseNot(tokens, parseValue));
-    }
-    return { and: list };
+export function parseAnd<T>(parseValue: (cursor: Cursor) => Expr.Value<T>): (tokens: Cursor) => Expr.And<T> {
+    return (tokens: Cursor) => {
+        const list = [parseNot(parseValue)(tokens)];
+        while (tokens.peekToken()?.type === "logicalAnd") {
+            tokens.nextToken(); // Consume the `&`
+            list.push(parseNot(parseValue)(tokens));
+        }
+        return { and: list };
+    };
 }
 
 /** Parse a {@link Expr.Not} using a `parseValue` to parse the potentially negated value */
-function parseNot<T>(tokens: Cursor, parseValue: (cursor: Cursor) => Expr.Value<T>): Expr.Not<T> {
-    let not = false;
-    if (tokens.peekToken()?.type === "logicalNot") {
-        tokens.nextToken(); // Consume the `!`
-        not = true;
-    }
-    return { not, value: parseValue(tokens) };
+export function parseNot<T>(parseValue: (cursor: Cursor) => Expr.Value<T>): (tokens: Cursor) => Expr.Not<T> {
```

Comment:
why does this return a lambda?
Isn't it only ever called in `parseAnd` which immediately calls it exactly once?